firmware/eos.c

```diff
@@ -0,0 +1,476 @@
+    return inited;
+}
+
+bool eos_signingIsFinished(void) {
+    return inited && actions_remaining == 0 && unknown_remaining == 0;
+}
+
+uint32_t eos_actionsRemaining(void) {
+    return actions_remaining;
+}
+
+void eos_signingAbort(void) {
+    inited = false;
+    memzero(&hasher_preimage, sizeof(hasher_preimage));
+    memzero(&hasher_unknown, sizeof(hasher_unknown));
+    memzero(&header, sizeof(header));
+    memzero(&node, sizeof(node));
+    memzero(&root, sizeof(root));
+    memzero(address_n, sizeof(address_n));
+    address_n_count = 0;
+    actions_remaining = 0;
+    unknown_remaining = 0;
+    unknown_total = 0;
+}
+
+bool eos_compileAsset(const EosAsset *asset) {
+    if (!asset->has_amount)
+        return false;
+
+    if (!asset->has_symbol)
+        return false;
+
+    hasher_Update(&hasher_preimage, (const uint8_t*)&asset->amount, 8);
+    hasher_Update(&hasher_preimage, (const uint8_t*)&asset->symbol, 8);
+    return true;
+}
+
+bool eos_compileString(const char *str) {
+    if (!str)
+        return false;
+    uint32_t len = strlen(str);
+    eos_hashUInt(&hasher_preimage, len);
+    if (len)
+        hasher_Update(&hasher_preimage, (const uint8_t*)str, len);
+    return true;
+}
+
+bool eos_compileActionCommon(const EosActionCommon *common) {
+    if (!(actions_remaining--))
+        return false;
+
+    if (!common->has_account)
+        return false;
+
+    if (!common->has_name)
+        return false;
+
+    if (!common->authorization_count)
+        return false;
+
+    hasher_Update(&hasher_preimage, (const uint8_t*)&common->account, 8);
+    hasher_Update(&hasher_preimage, (const uint8_t*)&common->name, 8);
+
+    eos_hashUInt(&hasher_preimage, common->authorization_count);
+    for (size_t i = 0; i < common->authorization_count; i++) {
+        if (!eos_compilePermissionLevel(&common->authorization[i]))
+            return false;
+    }
+
+    return true;
+}
+
+bool eos_compilePermissionLevel(const EosPermissionLevel *auth) {
+    if (!auth->has_actor)
+        return false;
+
+    if (!auth->has_permission)
+        return false;
+
+    hasher_Update(&hasher_preimage, (const uint8_t*)&auth->actor, 8);
+    hasher_Update(&hasher_preimage, (const uint8_t*)&auth->permission, 8);
+
+    return true;
+}
+
+bool eos_hasActionUnknownDataRemaining(void) {
+    return 0 < unknown_remaining;
+}
+
+static bool isSupportedAction(const EosActionCommon *common) {
+    if (common->account == EOS_eosio ||
+        common->account == EOS_eosio_token) {
+        switch (common->name) {
+        case EOS_Transfer:
+        case EOS_Owner:
+        case EOS_Active:
+        case EOS_DelegateBW:
+        case EOS_UndelegateBW:
+        case EOS_Refund:
+        case EOS_BuyRam:
+        case EOS_BuyRamBytes:
+        case EOS_SellRam:
+        case EOS_VoteProducer:
+        case EOS_UpdateAuth:
+        case EOS_DeleteAuth:
+        case EOS_LinkAuth:
+        case EOS_UnlinkAuth:
+            return true;
+        }
```

Comment:
Note to self: add `__builtin_unreachable()` to force enum additions to also be added to the switch.